pkg/cloudlogging/puller.go

```diff
@@ -111,3 +113,44 @@ func (p *Puller) Pull(ctx context.Context, filter string, maxCount int) ([]*logg
 	}
 	return ls, nil
 }
+
+// SteamPull pulls live log entries, it won't stop until a cancel signal happens.
+func (p *Puller) StreamPull(ctx context.Context, filter string, logCh chan<- *loggingpb.LogEntry) error {
+	var tailLogEntriesClient loggingpb.LoggingServiceV2_TailLogEntriesClient
+	var createClientErr error
+	if tailLogEntriesClient, createClientErr = p.client.TailLogEntries(ctx); createClientErr != nil {
+		return fmt.Errorf("failed to create tailLogEntriesClient: %w", createClientErr)
+	}
+	defer func() {
+		_ = tailLogEntriesClient.CloseSend()
+	}()
+
+	if err := retry.Do(ctx, p.retry, func(ctx context.Context) error {
+		req := &loggingpb.TailLogEntriesRequest{
+			ResourceNames: []string{p.resource},
+			Filter:        filter,
+		}
+
+		if err := tailLogEntriesClient.Send(req); err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to send request: %w", err))
+		}
+
+		for {
+			resp, err := tailLogEntriesClient.Recv()
+			if errors.Is(err, io.EOF) {
+				continue
+			}
+			if err != nil {
+				return retry.RetryableError(fmt.Errorf("failed to receive response: %w", err))
+			}
+			if resp.GetEntries() != nil {
+				for _, v := range resp.GetEntries() {
+					logCh <- v
+				}
+			}
+		}
+	}); err != nil {
+		return fmt.Errorf("failed to pull log entries: %w", err)
+	}
+	return nil
+}
```

Comment:
```suggestion
func (p *Puller) StreamPull(ctx context.Context, filter string, logCh chan<- *loggingpb.LogEntry) (err error) {
	tailClient, err := p.client.TailLogEntries(ctx)
	if err != nil {
		return fmt.Errorf("failed to create tailLogEntriesClient: %w", err)
	}
	defer func() {
		err = tailClient.CloseSend()
	}()

	if err := retry.Do(ctx, p.retry, func(ctx context.Context) error {
		req := &loggingpb.TailLogEntriesRequest{
			ResourceNames: []string{p.resource},
			Filter:        filter,
		}

		if err := tailLogEntriesClient.Send(req); err != nil {
			return retry.RetryableError(fmt.Errorf("failed to send request: %w", err))
		}

		for {
			resp, err := tailLogEntriesClient.Recv()
			if errors.Is(err, io.EOF) {
				continue
			}
			if err != nil {
				return retry.RetryableError(fmt.Errorf("failed to receive response: %w", err))
			}
			if resp.GetEntries() != nil {
				for _, v := range resp.GetEntries() {
					logCh <- v
				}
			}
		}
	}); err != nil {
		return fmt.Errorf("failed to pull log entries: %w", err)
	}
	return nil
}
```

Reply:
Shouldn't ignore error in the deferred func

Reply:
Updated